src/aspire/classification/rir_class2d.py

```diff
@@ -0,0 +1,577 @@
+                f"RIRClass2D basis.dtype {self.fb_basis.dtype} does not match self.dtype {self.dtype}."
+            )
+
+        # Sanity Checks
+        assert hasattr(self.pca_basis, "calculate_bispectrum")
+
+        if self.src.n < self.bispectrum_componenents:
+            raise RuntimeError(
+                f"{self.src.n} Images too small for Bispectrum Componenents {self.bispectrum_componenents}."
+                "  Increase number of images or reduce components."
+            )
+
+        # Implementation Checks
+        # # Do we have a sane Nearest Neighbor
+        nn_implementations = {
+            "legacy": self._legacy_nn_classification,
+            "sklearn": self._sk_nn_classification,
+        }
+        if nn_implementation not in nn_implementations:
+            raise ValueError(
+                f"Provided nn_implementation={nn_implementation} not in {nn_implementations.keys()}"
+            )
+        self._nn_classification = nn_implementations[nn_implementation]
+
+        # # Do we have a sane Large Dataset PCA
+        large_pca_implementations = {
+            "legacy": self._legacy_pca,
+            "sklearn": self._sk_pca,
+        }
+        if large_pca_implementation not in large_pca_implementations:
+            raise ValueError(
+                f"Provided large_pca_implementation={large_pca_implementation} not in {large_pca_implementations.keys()}"
+            )
+        self._pca = large_pca_implementations[large_pca_implementation]
+
+        # # Do we have a sane Bispectrum component
+        bispectrum_implementations = {
+            "legacy": self._legacy_bispectrum,
+            "devel": self._devel_bispectrum,
+        }
+        if bispectrum_implementation not in bispectrum_implementations:
+            raise ValueError(
+                f"Provided bispectrum_implementation={bispectrum_implementation} not in {bispectrum_implementations.keys()}"
+            )
+        elif bispectrum_implementation == "legacy" and self._pca != self._legacy_pca:
+            raise ValueError(
+                '"legacy" bispectrum_implementation implies'
+                ' large_pca_implementation="legacy".'
+                " Check class configuration and retry."
+            )
+        self._bispectrum = bispectrum_implementations[bispectrum_implementation]
+
+    def classify(self):
+        """
+        This is the high level method to perform the 2D images classification.
+
+        The stages of this method are intentionally modular so they may be
+        swapped for other implementations.
+        """
+
+        # # Stage 1: Compute coef and reduce dimensionality.
+        # Memioze/batch this later when result is working
+        # Initial round of component truncation is before bispectrum.
+        #  default of 400 components was taken from legacy code.
+        # Instantiate a new compressed (truncated) basis.
+        self.pca_basis = self.pca_basis.compress(self.fspca_components)
```

Comment:
I'm wondering if the FSPCA basis should be constructed here as well. Right now, half of the FSPCA processing is happening before the `RIRClass2D` instantiation and half is happening here.

Reply:
Right, I see two separate questions relating to this.

1) Should `FSPCABasis` just be compressed when instantiated?  I kept it flexible, in part for debugging.  It can be consolidated if we simply admit the number of compresses components in `__init__()`.
2) In `RIRClass2D` should `pca_basis` be an optional arg with default to construct an FSPCABasis?  This is relatively simple.  Besides the obvious couple lines of code, it would require `RIRClass2D` to optionally take args for `pca_basis`.  Presently only `noise_var`.  We could admit a dict of kwargs `pca_basis_kwargs` or something so it's extensible later.

Both of those are easy... Just let me know whether they are good by you and I'll make the changes.

Reply:
>     1. Should `FSPCABasis` just be compressed when instantiated?  I kept it flexible, in part for debugging.  It can be consolidated if we simply admit the number of compresses components in `__init__()`.

I think so. How does the `FSPCABasis` act uncompressed? It just reproduces the parent basis? If so, it should be compressed from the beginning.

>     2. In `RIRClass2D` should `pca_basis` be an optional arg with default to construct an FSPCABasis?  This is relatively simple.  Besides the obvious couple lines of code, it would require `RIRClass2D` to optionally take args for `pca_basis`.  Presently only `noise_var`.  We could admit a dict of kwargs `pca_basis_kwargs` or something so it's extensible later.

Right I didn't think of that. I would still lean towards including it in `RIRClass2D` instantiation. The other option would be to completely decouple `FSPCABasis` and `RIRClass2D`. In other words, you just feed it coefficients in any (steerable) basis and it doesn't care if it's FSPCA or not. If I've understood the code correctly, this should just be a matter of removing the `compress` call (or including it in FSPCA instantiation, as you suggest).

Reply:
(2) Well technically we already have two other meaningful classes, `FBBasis2D` and `FFBBasis2D`. It's a very expensive bispectrum calculation, but that's another matter. Think of it this way, if a user (or developer) wants to test the performance of the algorithm without compression, why should they go through the trouble of constructing a PCA basis (and keeping all the components)? It's just another moving part that doesn't *have* to be included in the `RIRClass2D`.

Reply:
Looking now, it seems one of the complications of auto building the pca_basis is how we want to plumb through the (fb basis) coef if they are already computed.   I guess the instantiator can admit an optional `coef`.  Previously I had it as an optional argument of `build`.

Reply:
Changed to automatically `_compress` as part of `build`, which is now part of the initialization (of FSPCA).

I changed the RIR class to admit an optional basis arg.  When not provided it will generate an FSPCA compressed to the target number of `fscpa_components` during `classify`. 

Similarly, `basis` and `coef` can be provided to `FSPCABasis` as an optimization for users that already have FB coefs and want to avoid recomputing them.

This required some minor changes to the unit tests and notebook.  Will push later today (have to run through a few meetings).

Reply:
Is supplying cached FB coefficients a must-have feature? Given that we'll probably be batching this at some point, it seems an unlikely use case.

Reply:
Removed the optional `coef`.